src/centralized.rs

```diff
@@ -0,0 +1,499 @@
+    pub fn new(v: usize, t: usize, n_p: usize) -> Round {
+        let mut data = vec![v.to_ne_bytes().to_vec()];
+        data.push(t.to_ne_bytes().to_vec());
+        let (h, h_usize) = Round::h1(data, n_p);
+        Round {
+            v,
+            t,
+            s_list: vec![],
+            h,
+            h_usize,
+            n_p,
+        }
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Round, s: Data) -> Round {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let mut data = vec![r.h.clone().to_vec()];
+        data.push(s.to_vec());
+        let (h, h_usize) = Round::h1(data, r.n_p);
+        Round {
+            v: r.v,
+            t: r.t,
+            s_list,
+            h,
+            h_usize,
+            n_p: r.n_p,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
+    /// Proof counter
+    r: usize,
+    /// Proof 2nd counter
+    d: usize,
+    /// Proof tuple
+    items: Vec<Data>,
+}
+
+impl Proof {
+    /// Returns a new proof
+    fn new() -> Self {
+        Proof {
+            r: 0,
+            d: 0,
+            items: vec![],
+        }
+    }
+
+    /// Oracle producing a uniformly random value in [1, n_p] used for prehashing S_p
+    fn h0(setup: &Setup, v: usize, s: Data) -> usize {
+        let mut data = vec![v.to_ne_bytes().to_vec()];
+        data.push(s.to_vec());
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(data);
+        utils::oracle_uniform(&digest, setup.n_p)
+    }
+
+    /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0 otherwise
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut data = vec![];
+        data.push(r.v.to_ne_bytes().to_vec());
+        data.push(r.t.to_ne_bytes().to_vec());
+        for s in &r.s_list {
+            data.push(s.clone().to_vec());
+        }
+        let digest = utils::combine_hashes::<DIGEST_SIZE>(data);
+        utils::oracle_binomial(&digest, setup.q)
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(
+        setup: &Setup,
+        bins: &Vec<Vec<Data>>,
+        round: &Round,
+        nb_steps: Arc<AtomicUsize>,
+    ) -> Option<Proof> {
+        if round.s_list.len() == setup.u {
+            if Proof::h2(setup, round) {
+                let r = round.v;
+                let d = round.t;
+                let items = round.s_list.clone();
+                return Some(Proof { r, d, items });
+            } else {
+                return None;
+            }
+        }
+        let result = bins[round.h_usize].par_iter().find_map_first(|&s| {
```

Comment:
You can simplify the return statement by simply replacing lines 301--309 by 
```
    bins[round.h_usize]
        .par_iter()
        .find_map_first(|&s| {
            // Stop if nb_steps has reached the threshold
            if nb_steps.load(atomic::Ordering::Relaxed) == setup.d {
                return None;
            }

            // Increment the step counter
            nb_steps.fetch_add(1, atomic::Ordering::Relaxed);

            // Recursive DFS with updated round
            Self::dfs(setup, bins, &Round::update(round, s), Arc::clone(&nb_steps))
        })
```

Reply:
@djetchev As I removed the parallelization (to be treated in its own issue) I have updated the code similarly to your idea but with a fold.